File: lib/models/aqi_reading.dart
class AQIReading {
  final String id;
  final String deviceId;
  final String? sourceName; // e.g., "WAQI", "OpenWeather", "My Blue Device"
  final DateTime timestamp;
  final double pm25;
  final double? pm10;
  final double? no2;
  final double? co;
  final double? so2;
  final double? o3;
  final double? nh3;
  final double? temperature;
  final double? humidity;
  final double? latitude;
  final double? longitude;
  final String sourceType; // bluetooth, api, manual
  final String? contextTag; // indoor, outdoor, unknown
  final String syncStatus; // pending, synced, failed

  AQIReading({
    required this.id,
    required this.deviceId,
    this.sourceName,
    required this.timestamp,
    required this.pm25,
    this.pm10,
    this.no2,
    this.co,
    this.so2,
    this.o3,
    this.nh3,
    this.temperature,
    this.humidity,
    this.latitude,
    this.longitude,
    required this.sourceType,
    this.contextTag,
    this.syncStatus = 'pending',
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'device_id': deviceId,
      'source_name': sourceName,
      'timestamp': timestamp.toUtc().toIso8601String(),
      'pm25': pm25,
      'pm10': pm10,
      'no2': no2,
      'co': co,
      'so2': so2,
      'o3': o3,
      'nh3': nh3,
      'temperature': temperature,
      'humidity': humidity,
      'latitude': latitude,
      'longitude': longitude,
      'source_type': sourceType,
      'context_tag': contextTag,
      'sync_status': syncStatus,
    };
  }

  factory AQIReading.fromMap(Map<String, dynamic> map) {
    return AQIReading(
      id: map['id'],
      deviceId: map['device_id'],
      sourceName: map['source_name'],
      timestamp: DateTime.parse(map['timestamp']),
      pm25: (map['pm25'] as num).toDouble(),
      pm10: map['pm10'] != null ? (map['pm10'] as num).toDouble() : null,
      no2: map['no2'] != null ? (map['no2'] as num).toDouble() : null,
      co: map['co'] != null ? (map['co'] as num).toDouble() : null,
      so2: map['so2'] != null ? (map['so2'] as num).toDouble() : null,
      o3: map['o3'] != null ? (map['o3'] as num).toDouble() : null,
      nh3: map['nh3'] != null ? (map['nh3'] as num).toDouble() : null,
      temperature: map['temperature'] != null ? (map['temperature'] as num).toDouble() : null,
      humidity: map['humidity'] != null ? (map['humidity'] as num).toDouble() : null,
      latitude: map['latitude'] != null ? (map['latitude'] as num).toDouble() : null,
      longitude: map['longitude'] != null ? (map['longitude'] as num).toDouble() : null,
      sourceType: map['source_type'],
      contextTag: map['context_tag'],
      syncStatus: map['sync_status'] ?? 'pending',
    );
  }

  AQIReading copyWith({
    String? id,
    String? deviceId,
    String? sourceName,
    DateTime? timestamp,
    double? pm25,
    double? pm10,
    double? no2,
    double? co,
    double? so2,
    double? o3,
    double? nh3,
    double? temperature,
    double? humidity,
    double? latitude,
    double? longitude,
    String? sourceType,
    String? contextTag,
    String? syncStatus,
  }) {
    return AQIReading(
      id: id ?? this.id,
      deviceId: deviceId ?? this.deviceId,
      sourceName: sourceName ?? this.sourceName,
      timestamp: timestamp ?? this.timestamp,
      pm25: pm25 ?? this.pm25,
      pm10: pm10 ?? this.pm10,
      no2: no2 ?? this.no2,
      co: co ?? this.co,
      so2: so2 ?? this.so2,
      o3: o3 ?? this.o3,
      nh3: nh3 ?? this.nh3,
      temperature: temperature ?? this.temperature,
      humidity: humidity ?? this.humidity,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      sourceType: sourceType ?? this.sourceType,
      contextTag: contextTag ?? this.contextTag,
      syncStatus: syncStatus ?? this.syncStatus,
    );
  }
}

File: rn_aqi_app/src/models/AqiReading.ts
export type SourceType = 'bluetooth' | 'api' | 'manual';

export interface AQIReading {
  id: string;
  deviceId: string;
  timestamp: string; // ISO-8601
  pm25: number;
  pm10?: number;
  o3?: number;
  no2?: number;
  so2?: number;
  co?: number;
  sourceType: SourceType;
  latitude?: number;
  longitude?: number;
  contextTag?: string; // 'indoor' | 'outdoor' | 'unknown'
  stationMetadata?: {
    manufacturer?: string;
    model?: string;
    owner?: string;
  };
}

export type AQILevel = 'Good' | 'Moderate' | 'Unhealthy (Sens.)' | 'Unhealthy' | 'Very Unhealthy' | 'Hazardous' | 'Connecting...';

export const getAQILevel = (pm25: number): { level: AQILevel; color: string } => {
  if (pm25 === -1) return { level: 'Connecting...', color: '#9E9E9E' };
  if (pm25 <= 12.0) return { level: 'Good', color: '#4CAF50' };
  if (pm25 <= 35.4) return { level: 'Moderate', color: '#FFEB3B' };
  if (pm25 <= 55.4) return { level: 'Unhealthy (Sens.)', color: '#FF9800' };
  if (pm25 <= 150.4) return { level: 'Unhealthy', color: '#F44336' };
  if (pm25 <= 250.4) return { level: 'Very Unhealthy', color: '#9C27B0' };
  return { level: 'Hazardous', color: '#7E0023' };
};

export const getHealthRecommendation = (level: AQILevel): string => {
  switch (level) {
    case 'Good':
      return 'Air quality is satisfactory, and air pollution poses little or no risk.';
    case 'Moderate':
      return 'Air quality is acceptable. However, there may be a risk for some people, particularly those who are unusually sensitive to air pollution.';
    case 'Unhealthy (Sens.)':
      return 'Members of sensitive groups may experience health effects. The general public is less likely to be affected.';
    case 'Unhealthy':
      return 'Some members of the general public may experience health effects; members of sensitive groups may experience more serious health effects.';
    case 'Very Unhealthy':
      return 'Health alert: The risk of health effects is increased for everyone.';
    case 'Hazardous':
      return 'Health warning of emergency conditions: everyone is more likely to be affected.';
    default:
      return 'Retrieving real-time data...';
  }
};

/**
 * Calculates the exact US AQI score (0-500) from a raw PM2.5 concentration (µg/m³).
 * Uses the standard US EPA linear interpolation formula.
 */
export const calculateAQI = (pm25: number): number => {
  if (pm25 < 0) return 0;
  if (pm25 <= 12.0) return Math.round(((50 - 0) / (12.0 - 0.0)) * (pm25 - 0.0) + 0);
  if (pm25 <= 35.4) return Math.round(((100 - 51) / (35.4 - 12.1)) * (pm25 - 12.1) + 51);
  if (pm25 <= 55.4) return Math.round(((150 - 101) / (55.4 - 35.5)) * (pm25 - 35.5) + 101);
  if (pm25 <= 150.4) return Math.round(((200 - 151) / (150.4 - 55.5)) * (pm25 - 55.5) + 151);
  if (pm25 <= 250.4) return Math.round(((300 - 201) / (250.4 - 150.5)) * (pm25 - 150.5) + 201);
  if (pm25 <= 350.4) return Math.round(((400 - 301) / (350.4 - 250.5)) * (pm25 - 250.5) + 301);
  if (pm25 <= 500.4) return Math.round(((500 - 401) / (500.4 - 350.5)) * (pm25 - 350.5) + 401);
  return 500;
};
